### PR TITLE
Deprecate `Trainer.devices` in favor of `Trainer.num_devices` and  `Trainer.device_ids` 

### PR DESCRIPTION
**Pull request by DuYicong515**
## What does this PR do?
Part of #12171 

### Does your PR introduce any breaking changes? If yes, please list them.

N/A

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @four4fish  @awaelchli

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -2026,6 +2026,21 @@ def should_rank_save_checkpoint(self) -> bool:
...
+        for idx, device in enumerate(devices):
+            if isinstance(device, torch.device):
+                device_ids.append(device.index or idx)
+            elif isinstance(device, int):
+                device_ids.append(device)
+        return device_ids
+
+    @property
+    def num_devices(self) -> int:
+        return len(self.device_ids)
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2042]: can it rely on `trainer._accelerator_connector.devices` here? since we are only concerned with the count, and not the actual devices. `device_ids` does some additional minor computation.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -2048,7 +2063,11 @@ def num_gpus(self) -> int:
 
     @property
     def devices(self) -> Optional[Union[List[int], str, int]]:
-        return self._accelerator_connector.devices
+        rank_zero_deprecation(
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:2066]: can you add a deprecation test?

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (DuYicong515) [reply pytorch_lightning/trainer/trainer.py:2042]: I am planning to remove devices from `AcceleratorConnector` since that's not internally used by `AcceleratorConnector` and it's not meant to be a public property. 

But I could definitely implement the same logic here so the computations are more efficient.

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -2026,6 +2026,21 @@ def should_rank_save_checkpoint(self) -> bool:
...
+        device_ids = []
+        for idx, device in enumerate(devices):
+            if isinstance(device, torch.device):
+                device_ids.append(device.index or idx)
+            elif isinstance(device, int):
+                device_ids.append(device)
+        return device_ids
+
+    @property
+    def num_devices(self) -> int:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2041]: Please add docstrings for these. The docstring should clarify that that these are the device_ids *per node*

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (DuYicong515) [reply pytorch_lightning/trainer/trainer.py:2072]: Added in `tests/trainer/test_trainer.py`

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (DuYicong515) [reply pytorch_lightning/trainer/trainer.py]: Added!

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (DuYicong515) [reply pytorch_lightning/trainer/trainer.py]: Moved the logic here. Also in tests checked they are the same with len(device_ids)

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -2146,3 +2146,26 @@ def test_dataloaders_are_not_loaded_if_disabled_through_limit_batches(running_st
...
+    [
+        ({}, [0]),
+        ({"devices": 1}, [0]),
+        ({"accelerator": "gpu", "devices": 1}, [0]),
+        ({"strategy": "ddp", "devices": 1}, [0]),
+        ({"strategy": "ddp", "accelerator": "gpu", "devices": 1}, [0]),
+        ({"strategy": "ddp", "devices": 2}, [0, 1]),
+        ({"strategy": "ddp", "accelerator": "gpu", "devices": 2}, [0, 1]),
+        ({"strategy": "ddp", "accelerator": "gpu", "devices": [2]}, [2]),
+        ({"strategy": "ddp", "accelerator": "gpu", "devices": [0, 2]}, [0, 2]),
```
Comment (kaushikb11) [inline tests/trainer/test_trainer.py:2162]: It's not required to have `strategy` as a parameter for the test

File: tests/deprecated_api/test_remove_1-8.py [hunk 5]
```diff
@@ -688,3 +688,29 @@ def on_load_checkpoint(self, checkpoint):
...
+    [
+        ({}, 1),
+        ({"devices": 1}, 1),
+        ({"accelerator": "gpu", "devices": 1}, 1),
+        ({"strategy": "ddp", "devices": 1}, 1),
+        ({"strategy": "ddp", "accelerator": "gpu", "devices": 1}, 1),
+        ({"strategy": "ddp", "devices": 2}, 2),
+        ({"strategy": "ddp", "accelerator": "gpu", "devices": 2}, 2),
+        ({"strategy": "ddp", "accelerator": "gpu", "devices": [2]}, 1),
+        ({"strategy": "ddp", "accelerator": "gpu", "devices": [0, 2]}, 2),
```
Comment (kaushikb11) [inline tests/deprecated_api/test_remove_1-8.py:704]: Same here

**Review by kaushikb11 — APPROVED** [review]

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -2146,3 +2146,26 @@ def test_dataloaders_are_not_loaded_if_disabled_through_limit_batches(running_st
...
+@pytest.mark.parametrize(
+    ["trainer_kwargs", "expected_device_ids"],
+    [
+        ({}, [0]),
+        ({"devices": 1}, [0]),
+        ({"accelerator": "gpu", "devices": 1}, [0]),
+        ({"strategy": "ddp", "devices": 1}, [0]),
+        ({"strategy": "ddp", "accelerator": "gpu", "devices": 1}, [0]),
+        ({"strategy": "ddp", "devices": 2}, [0, 1]),
+        ({"strategy": "ddp", "accelerator": "gpu", "devices": 2}, [0, 1]),
```
Comment (kaushikb11) [inline tests/trainer/test_trainer.py:2160]: Could add set of kwargs with strings being passed to `devices`

File: tests/trainer/test_trainer.py [hunk 7]
```diff
@@ -2146,3 +2146,26 @@ def test_dataloaders_are_not_loaded_if_disabled_through_limit_batches(running_st
...
     )
     assert dl is None
+
+
+@pytest.mark.parametrize(
+    ["trainer_kwargs", "expected_device_ids"],
+    [
+        ({}, [0]),
+        ({"devices": 1}, [0]),
+        ({"accelerator": "gpu", "devices": 1}, [0]),
```
Comment (kaushikb11) [inline tests/trainer/test_trainer.py:2156]: Please add tests for TPUs and IPUs, you could mock the availability

File: tests/trainer/test_trainer.py [hunk 8]
```diff
@@ -2146,3 +2146,26 @@ def test_dataloaders_are_not_loaded_if_disabled_through_limit_batches(running_st
         else getattr(trainer, f"{dl_prefix}_dataloaders")
     )
     assert dl is None
+
+
+@pytest.mark.parametrize(
+    ["trainer_kwargs", "expected_device_ids"],
```
Comment (four4fish) [inline tests/trainer/test_trainer.py:2152]: Can we test num_devices as well please?

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -2026,6 +2026,27 @@ def should_rank_save_checkpoint(self) -> bool:
     def num_nodes(self) -> int:
         return getattr(self.strategy, "num_nodes", 1)
 
+    @property
+    def device_ids(self) -> List[int]:
+        """List of device indexes per node."""
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/trainer.py:2031]:
```suggestion
        """List of device indexes"""
```

File: tests/trainer/test_trainer.py (hunk 8 above)
Comment (DuYicong515) [reply tests/trainer/test_trainer.py:2152]: assert len(trainer.device_ids) == trainer.num_devices should already test it. See L2171

Comment (four4fish) [reply tests/trainer/test_trainer.py:2152]: Oh  I see! nice!

File: pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -2026,6 +2026,27 @@ def should_rank_save_checkpoint(self) -> bool:
...
+        return device_ids
+
+    @property
+    def num_devices(self) -> int:
+        """Number of devices per node."""
+        if isinstance(self.strategy, SingleDeviceStrategy):
+            return 1
+        elif isinstance(self.strategy, ParallelStrategy):
+            return len(self.strategy.parallel_devices)
+        return 0
```
Comment (ethanwharris) [inline pytorch_lightning/trainer/trainer.py:2048]: Is this line reachable? Doesn't seem to come up in tests. Might be cleaner to just return `len(parallel_devices)` for parallel strategy or 1 otherwise

**Review by ethanwharris — APPROVED** [review]
LGTM, small comment

File: pytorch_lightning/trainer/trainer.py [hunk 11]
```diff
@@ -2026,6 +2026,22 @@ def should_rank_save_checkpoint(self) -> bool:
...
+        for idx, device in enumerate(devices):
+            if isinstance(device, torch.device):
+                device_ids.append(device.index or idx)
+            elif isinstance(device, int):
+                device_ids.append(device)
+        return device_ids
+
+    @property
+    def num_devices(self) -> int:
+        return len(self.strategy.parallel_devices) if isinstance(self.strategy, ParallelStrategy) else 1
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2043]: did the docstrings get lost? please clarify that these properties are per node
```suggestion
    def device_ids(self) -> List[int]:
        """List of device indexes used per node."""
        devices = getattr(self.strategy, "parallel_devices", [self.strategy.root_device])
        device_ids = []
        for idx, device in enumerate(devices):
            if isinstance(device, torch.device):
                device_ids.append(device.index or idx)
            elif isinstance(device, int):
                device_ids.append(device)
        return device_ids

    @property
    def num_devices(self) -> int:
        """ Number of devices the trainer uses per node. """
        return len(self.strategy.parallel_devices) if isinstance(self.strategy, ParallelStrategy) else 1
```

File: tests/trainer/test_trainer.py (hunk 7 above)
Comment (DuYicong515) [reply tests/trainer/test_trainer.py:2160]: The TPU have a lot of things to mock. It uses functions like `xm.xla_device` frequently not only we have to mock `_TPU_AVAILABLE`. I put more test cases in test_tpu.py and only run when we have tpus available. Should be similar for testing purposes. 

let me know what you think!

 https://github.com/PyTorchLightning/pytorch-lightning/blob/15364c18c81909e213f1910a57982f100f8a356d/pytorch_lightning/strategies/single_tpu.py#L44

File: pytorch_lightning/trainer/trainer.py (hunk 11 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:2043]: also shouldn't these properties be kept in sync? why isn't `num_devices` implemented like this?
```py
    return max(len(self.device_ids), 1)
```
?

File: pytorch_lightning/trainer/trainer.py (hunk 10 above)
Comment (DuYicong515) [reply pytorch_lightning/trainer/trainer.py]: Changed! Thanks for the advice.

File: pytorch_lightning/trainer/trainer.py (hunk 11 above)
Comment (DuYicong515) [reply pytorch_lightning/trainer/trainer.py:2043]: I received a comment previously about `self.device_ids` requiring additional computations. 
https://github.com/PyTorchLightning/pytorch-lightning/pull/12151#discussion_r817486457

I also add a test case to asserting `self.num_devices == len(expected_device_ids) == len(self.device_ids)`

I'm okay with both weighing more towards logic in sync or efficient computations.

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (DuYicong515) [reply pytorch_lightning/trainer/trainer.py:2031]: I feel per node is more clear? Thoughts? cc @ananthsub

File: tests/deprecated_api/test_remove_1-8.py [hunk 12]
```diff
@@ -688,3 +689,38 @@ def on_load_checkpoint(self, checkpoint):
         " v1.6 and will be removed in v1.8. Use `load_state_dict` instead."
     ):
         trainer.fit(model)
+
+
+@pytest.mark.parametrize(
```
Comment (rohitgr7) [inline tests/deprecated_api/test_remove_1-8.py:694]: I don't think all of these are required here since they are already tested in other files. Just take one of these cases and check the deprecated_call as you have already done at the end.

File: tests/trainer/test_trainer.py (hunk 7 above)
Comment (rohitgr7) [reply tests/trainer/test_trainer.py:2160]: I agree! for TPU it might not be that trivial. You can test it on TPU directly.

**Review by four4fish — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:2031]: I agree. This property is ambiguously named, so the comments should clarify that it's per node, and not the global list of device indices being used.

File: pytorch_lightning/trainer/trainer.py [hunk 13]
```diff
@@ -2026,6 +2026,23 @@ def should_rank_save_checkpoint(self) -> bool:
...
+            if isinstance(device, torch.device):
+                device_ids.append(device.index or idx)
+            elif isinstance(device, int):
+                device_ids.append(device)
+        return device_ids
+
+    @property
+    def num_devices(self) -> int:
+        """Number of devices the trainer uses per node."""
+        return len(self.strategy.parallel_devices) if isinstance(self.strategy, ParallelStrategy) else 1
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2044]: my assumption is that `num_devices` will be infrequently accessed, so I would prefer we rely on `device_ids` for the length to avoid these falling out of sync. If we need to optimize the performance, we can always do so without it being a breaking change.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 14]
```diff
@@ -2010,6 +2010,23 @@ def should_rank_save_checkpoint(self) -> bool:
     def num_nodes(self) -> int:
         return getattr(self.strategy, "num_nodes", 1)
 
+    @property
+    def device_ids(self) -> List[int]:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:2014]: I would have preferred if this was implemented on the accelerator connector. It is IMO not the responsibility of the Trainer to compute the device_ids off the strategy. Note the the strategy is owned by the connector, not the trainer. As the trainer has already many responsibilities, we should delegate as much as possible to the concrete components.

**Review by awaelchli — COMMENTED** [review]
Apologies for the delayed review.

File: pytorch_lightning/trainer/trainer.py (hunk 14 above)
Comment (DuYicong515) [reply pytorch_lightning/trainer/trainer.py:2014]: Thanks for reviewing @awaelchli!

My personal thoughts: the accelerator connector should stay internal, and accelerator connector should not expose those device related public properties, since they are easily computed off the strategy. 

> Note the the strategy is owned by the connector, not the trainer. 

Personally I feel the accelerator connector does a lot of checks/set up/initialization and configuration of the accelerator and strategy. After things are set up already, strategy/accelerator are already accessible on the Trainer. 

I think delegating computations using accelerator/strategy to the accelerator_connector will make us add many public properties to the internal-facing accelerator_connector. Deriving directly from strategy/accelerator directly are more straight forward, and they(strategy, accelerator) are very external facing as well. 

I'm very new to this and my feelings might be wrong. Feel free to add more thoughts to issue #12171!
